File: core/src/main/java/com/gumse/gui/XML/XMLLoader.cpp
#include "XMLLoader.h"
#include <System/Output.h>
#include <Essentials/Tools.h>
#include <Essentials/Filesystem/XMLReader.h>

#include "../Basics/Scroller.h"
#include "../Basics/Switch.h"
#include "../Basics/Button.h"
#include "../Basics/Slider.h"
#include "../Basics/Scroller.h"
#include "../Basics/TextBox.h"
#include "../Basics/Tabs.h"
#include "../Basics/Group.h"
#include "../Basics/Dropdown.h"
#include "../Primitives/RenderGUI.h"


XMLLoader::XMLLoader(std::string filename)
{
    pRootGUI = new RenderGUI();
    
    XMLReader(filename, XMLNode::NODE_TYPES::ELEMENT, [this](XMLNode* node) {
        std::string type = node->name;
        std::string content = node->content;
        RenderGUI* gui = nullptr;
        RenderGUI* parentGUI = nullptr;
        if(node->parent != nullptr)
            parentGUI = (RenderGUI*)node->parent->userptr;

        if(type == "script")
        {
            if(content != "")
                {//GUIScript* pScript = new GUIScript(content, pRootGUI);
                }
            else
                Gum::Output::error("XMLGUILoader: script is empty!");
        }
        else if(type == "style")
        {
            if(content != "")
                {//GUIStyle* pStyle = new GUIStyle(content, pRootGUI);
                }
            else
                Gum::Output::error("XMLGUILoader: style is empty!");
        }
        else if(type == "tab")
        {
            /*if(parentGUI->getType() == "Tabs")
            {
                RenderGUI* tabHandle = new RenderGUI();
                tabHandle->setSize(ivec2(100, 100));
                tabHandle->setSizeInPercent(true, true);
                tabHandle->setType("tab");
                xmlFileRecursiveFunc(node->children, tabHandle);
                std::string tabName = attrs["name"] != "" 
                                        ? attrs["name"] 
                                        : std::to_string(((Tabs*)parentGUI)->numTabs());
                bool active = attrs["active"] == "true";
                ((Tabs*)parentGUI)->addTab(tabName, active);
                ((Tabs*)parentGUI)->addGUIToTab(tabHandle, tabName);
            }
            else
            {
                Gum::Output::error("Tab tag added outside of tabs tag");
            }*/
        }
        else if(type == "item")
        {
            /*if(parentGUI->getType() == "Dropdown")
            {
                bool active = attrs["active"] == "true";
                ((Dropdown*)parentGUI)->addEntry(content, nullptr, active);
            }
            else
            {
                Gum::Output::error("item tag added outside of dropdown tag");
            }*/
        }
        else if(type == "gui-root") { gui = pRootGUI; }
        else if(type == "box")      { gui = Box::createFromXMLNode(node); }
        else if(type == "button")   { gui = Button::createFromXMLNode(node); }
        else if(type == "tabs")     { gui = Tabs::createFromXMLNode(node); }
        else if(type == "group")    { gui = Group::createFromXMLNode(node); }
        else if(type == "dropdown") { gui = Dropdown::createFromXMLNode(node); }
        else if(type == "scroller") { gui = Scroller::createFromXMLNode(node); }
        else if(type == "slider")   { gui = Slider::createFromXMLNode(node); }
        else if(type == "switch")   { gui = Switch::createFromXMLNode(node); }

        if(gui != nullptr)
        {
            processGeneralArguments(node->mAttributes, gui);
            node->userptr = gui;

            if(parentGUI != nullptr)
                parentGUI->addGUI(gui);
        }
    });
}

XMLLoader::~XMLLoader()
{

}


void XMLLoader::processGeneralArguments(std::map<std::string, std::string> args, RenderGUI* gui)
{
    for(auto arg : args)
    {
        std::string name = arg.first;
        std::vector<std::string> values = Tools::splitStr(arg.second, ',');
        
        if (name == "title")
        {
            gui->setTitle(values[0]);
        }
        else if(name == "pos")
        {
            gui->setPositionInPercent(Tools::strContains(values[0], "%"), Tools::strContains(values[1], "%"));
            gui->setPosition(Tools::StringToVec2(arg.second));
        }
        else if(name == "size")
        {
            gui->setSizeInPercent(Tools::strContains(values[0], "%"), Tools::strContains(values[1], "%"));
            gui->setSize(Tools::StringToVec2(arg.second));
        }
        else if(name == "origin")
        {
            gui->setOriginInPercent(Tools::strContains(values[0], "%"), Tools::strContains(values[1], "%"));
            gui->setOrigin(Tools::StringToVec2(arg.second));
        }
        else if(name == "margin")
        {
            gui->setMargin(Tools::StringToVec2(arg.second));
        }
        else if(name == "max-size")
        {
            gui->setMaxSizeInPercent(Tools::strContains(values[0], "%"), Tools::strContains(values[1], "%"));
            gui->setMaxSize(Tools::StringToVec2(arg.second));
        }
        else if(name == "min-size")
        {
            gui->setMinSizeInPercent(Tools::strContains(values[0], "%"), Tools::strContains(values[1], "%"));
            gui->setMinSize(Tools::StringToVec2(arg.second));
        }
        else if(name == "color")         { gui->setColor(vec4(Gum::Maths::HEXToRGB(arg.second.substr(1, arg.second.length())) / 255.0f, 1.0f)); }
        else if(name == "id")            { gui->setID(values[0]); }
        else if(name == "border-radius") { gui->setCornerRadius(Tools::StringToVec4(arg.second)); }
        else if(name == "tooltip")       { gui->setToolTip(values[0]); }
        else                             { gui->addAttribute(arg.first, arg.second); }
    }
}


//
// Getter
//
RenderGUI* XMLLoader::getRootGUI() { return this->pRootGUI; }

File: core/src/main/java/com/gumse/gui/XML/XMLLoader.java
package com.gumse.gui.XML;

import java.util.Map;

import com.gumse.gui.Basics.*;
import com.gumse.gui.Primitives.RenderGUI;
import com.gumse.gui.XML.XMLGUI.XMLGUICreator;
import com.gumse.maths.Color;
import com.gumse.maths.ivec2;
import com.gumse.maths.vec4;
import com.gumse.system.filesystem.XML.XMLNode;
import com.gumse.system.filesystem.XML.XMLReader;
import com.gumse.system.filesystem.XML.XMLReader.NodeCallback;
import com.gumse.tools.Output;
import com.gumse.tools.Toolbox;

public class XMLLoader
{
    private RenderGUI pRootGUI;
    
    public XMLLoader(String filename)
    {
        pRootGUI = new RenderGUI();
        pRootGUI.setType("XMLRoot");
        
        new XMLReader(filename, XMLNode.NODE_TYPES.ELEMENT, new NodeCallback() {
            @Override public void run(XMLNode node, int depth) 
            {
                String type = node.name;
                //String content = node.content;
                RenderGUI gui = null;
                RenderGUI parentGUI = null;
                if(node.parent != null)
                    parentGUI = (RenderGUI)node.parent.userptr;

                if(type.equals("item"))
                {
                    if(parentGUI.getType() == "Dropdown")
                    {
                        boolean active = node.getAttribute("active") == "true";
                        ((Dropdown)parentGUI).addEntry(node.content, null, active);
                    }
                    else
                    {
                        Output.error("item tag added outside of dropdown tag");
                    }
                }
                else if(type.equals("gui-root")) { gui = pRootGUI; }
                else if(type.equals("gui"))      { gui = new RenderGUI(); }
                else
                {
                    for(Map.Entry<String, XMLGUICreator> entry : XMLGUI.getTypes().entrySet())
                    {
                        if(type.equals(entry.getKey()))
                        {
                            gui = entry.getValue().create(node);
                            break;
                        }
                    }
                }

                if(gui != null)
                {
                    processGeneralArguments(node.mAttributes, gui);
                    node.userptr = gui;

                    if(parentGUI != null)
                        parentGUI.addGUI(gui);
                }
                
            }
        });
    }


    private void processGeneralArguments(Map<String, String> args, RenderGUI gui)
    {
        for(Map.Entry<String, String> arg : args.entrySet())
        {
            String name = arg.getKey();
            String valueStr = arg.getValue();
            String[] values = valueStr.split(",");
            
            if (name.equals("title"))
            {
                gui.setTitle(values[0]);
            }
            
            else if(name.equals("pos"))
            {
                gui.setPositionInPercent(values[0].contains("%"), values[1].contains("%"));
                gui.setPosition(new ivec2(Toolbox.StringToVec2(valueStr)));
            }
            else if(name.equals("size"))
            {
                gui.setSizeInPercent(values[0].contains("%"), values[1].contains("%"));
                gui.setSize(new ivec2(Toolbox.StringToVec2(valueStr)));
            }
            else if(name.equals("origin"))
            {
                gui.setOriginInPercent(values[0].contains("%"), values[1].contains("%"));
                gui.setOrigin(new ivec2(Toolbox.StringToVec2(valueStr)));
            }
            else if(name.equals("margin"))
            {
                gui.setMargin(new ivec2(Toolbox.StringToVec2(valueStr)));
            }
            else if(name.equals("max-size"))
            {
                gui.setMaxSizeInPercent(values[0].contains("%"), values[1].contains("%"));
                gui.setMaxSize(new ivec2(Toolbox.StringToVec2(valueStr)));
            }
            else if(name.equals("min-size"))
            {
                gui.setMinSizeInPercent(values[0].contains("%"), values[1].contains("%"));
                gui.setMinSize(new ivec2(Toolbox.StringToVec2(valueStr)));
            }
            else if(name.equals("color"))         { gui.setColor(vec4.div(Color.HEXToRGBA(valueStr), 255.0f)); }
            else if(name.equals("id"))            { gui.setID(new String(values[0])); }
            else if(name.equals("locale-id"))     { gui.setLocaleID(new String(values[0])); }
            else if(name.equals("border-radius")) { gui.setCornerRadius(Toolbox.StringToVec4(valueStr)); }
            else if(name.equals("tooltip"))       { gui.setToolTip(values[0]); }
            else                                  { gui.addAttribute(name, valueStr); }
        }
    }


    //
    // Getter
    //
    public RenderGUI getRootGUI() { return this.pRootGUI; }
};